File: servers/nodejs/tsp-solver.js
var express = require('express');
var app = express();
var expressWs = require('express-ws')(app);
var permutationEngine = require('permutation-engine');

app.use(express.static('public'));

app.listen(9999, function () {
  console.log('Server listening on port 9999!');
});

app.ws('/tsp', function(ws, req) {
  ws.on('message', function(msg) {
    var received_array = JSON.parse(msg);
    var coords_array = [];

    for(var i = 0; i < received_array.length; i += 2)
      {
      var node = [received_array[i],received_array[i+1]];
      coords_array.push(node);
      }

    if(coords_array.length < 3)
      {
      coords_array.unshift(0);
      ws.send(JSON.stringify(coords_array));
      return;
      }

    var perms_count = calculateFactorial(coords_array.length - 1);
    var permengine = new permutationEngine(coords_array.length - 1);
    var permutation;
    var shortest_permutation = [];
    var shortest_route_length = 1000000;
    var second_shortest_permutation;
    var second_shortest_route_length = 1000000;
    var route_length;
    var first_coords = coords_array.shift();

    for(var index = 0; index < perms_count; index++)
      {
      permutation = permengine.index2perm(index);
      route_length = getRouteLength(permutation,coords_array,first_coords);
      if(route_length < shortest_route_length)
        {
        second_shortest_route_length = shortest_route_length;
        second_shortest_permutation = shortest_permutation.slice();
        shortest_route_length = route_length;
        shortest_permutation = permutation;
        }
      else if(route_length < second_shortest_route_length && route_length > shortest_route_length)
        {
        second_shortest_route_length = route_length;
        second_shortest_permutation = permutation;
        }
      }

    var shortest_route_array = shortest_permutation.map(getCoordsByPermIndex,coords_array);
    shortest_route_array.unshift(first_coords); //beginning
    shortest_route_array.push(first_coords); //ending

    var second_shortest_route_array = [];
    var both_routes_array = [];
    if(second_shortest_route_length !== 1000000)
      {
      second_shortest_route_array = second_shortest_permutation.map(getCoordsByPermIndex,coords_array);
      second_shortest_route_array.unshift(first_coords); //beginning
      second_shortest_route_array.push(first_coords); //ending
      both_routes_array = shortest_route_array.concat(second_shortest_route_array);
      both_routes_array.unshift(1); //array to be returned contains both shortest and 2nd shortest paths
      ws.send(JSON.stringify(both_routes_array));
      return;
      }

    shortest_route_array.unshift(0); //array to be returned contains only shortest path
    ws.send(JSON.stringify(shortest_route_array));
  });
});


function calculateDistance(coords_1, coords_2)
{
  if(!coords_2) {return 0;}
  var delta_x = coords_1[0] - coords_2[0];
  var delta_y = coords_1[1] - coords_2[1];
  return Math.sqrt( Math.pow(delta_x, 2) + Math.pow(delta_y, 2) );
}

function calculateFactorial(num)
{
    var rval=1;
    for (var i = 2; i <= num; i++)
        rval = rval * i;
    return rval;
}

function getRouteLength(permutation,coords,first)
{
  var route_length = 0;

  for(var index = 0; index < permutation.length; index++)
    {
    route_length += calculateDistance(coords[permutation[index]-1], coords[permutation[index+1]-1]);
    }

  route_length += calculateDistance(first, coords[permutation[0]-1]);
  route_length += calculateDistance(first, coords[permutation[permutation.length-1]-1]);

  //truncate to four decimals because same route with different direction can have a small length difference
  return Math.round(Number(route_length) * 10000) / 10000;
}

function getCoordsByPermIndex( index )
{
  return this[index -1];
}

File: servers/python/tsp-solver.py
import tornado.httpserver
import tornado.websocket
import tornado.ioloop
import tornado.web
import json
import itertools
import math

class Coords:

    def __init__(self,coords,length):
        self.coords = coords
        self.length = float('%.7f' % length)

    def __eq__(self,other):
        index = -1

        for i in self.coords:
          if i[0] == other.coords[index][0] and i[1] == other.coords[index][1]:
            index-=1
          else:
            return False

        return True

    def __ne__(self,other):
        index = -1

        for i in self.coords:
          if i[0] == other.coords[index][0] and i[1] == other.coords[index][1]:
            return False
          else:
            index-=1

        return True

    def __lt__(self,other):
        if self.length < other.length:
          return True
        else:
          return False

    def __gt__(self,other):
        if self.length > other.length:
          return True
        else:
          return False

    def __hash__(self):
        return int(self.length * 1000000)

    def len(self):
      return self.length

    def route(self):
      return self.coords

def permutate(dataList):
  coord_list = []

  if(len(dataList) > 18):
    return "0"

  for i in range(0, len(dataList), 2):
    y_coord = dataList.pop()
    x_coord = dataList.pop()
    coord_list.append([x_coord,y_coord])

  coord_sub_list = coord_list[1:]

  perms = itertools.permutations(coord_sub_list)

  distances = []
  perms_list = []
  coords_obj_list = []

  for i in perms:
    perms_list.append(i)
    summa = calculate_distances(i)
    summa_2 = calculate_distances_to_first_node(i,coord_list)
    distances.append(summa+summa_2)
    coords_obj_list.append(Coords(i,summa+summa_2))

  coords_obj_list.sort()

  unique_list = list(set(coords_obj_list))
  unique_list.sort()

  min_path = []
  min_path_2 = []

  min_path = list(unique_list[0].route())
  min_path_length = unique_list[0].len()
  min_path.insert(0,coord_list[0])
  min_path.append(coord_list[0])

  if(len(unique_list) > 1):
    min_path_2 = list(unique_list[1].route())
    min_path_2_length = unique_list[1].len()
    min_path_2.insert(0,coord_list[0])
    min_path_2.append(coord_list[0])

  if(len(min_path_2) > 1):
    min_path += min_path_2
    min_path.insert(0,1) #array to be returned contains both shortest and 2nd shortest paths
  else:
    min_path.insert(0,0) #array to be returned contains only shortest path

  return json.dumps(min_path)

def calculate_distances(list_of_coords):
    sum = 0
    for x in range(1, len(list_of_coords)):
      sum += calculate_distance(list_of_coords[x-1],list_of_coords[x])
    return sum

def calculate_distance(lista1,lista2):
    delta_x = lista1[0] - lista2[0]
    delta_y = lista1[1] - lista2[1]
    return math.sqrt( math.pow(delta_x, 2) + math.pow(delta_y, 2))

def calculate_distances_to_first_node(route,coords):
    first_node = coords[0]
    route_head = route[0]
    route_tail = route[-1]
    from_first_node_to_head = calculate_distance(first_node,route_head)
    from_first_node_to_tail = calculate_distance(first_node,route_tail)
    sum = from_first_node_to_head + from_first_node_to_tail
    return sum

class WSHandler(tornado.websocket.WebSocketHandler):

  def check_origin(self, origin):
    return True

  def open(self):
    print('User is connected.')

  def on_message(self, message):
    obj = json.loads(message)
    self.write_message(permutate(obj))

  def on_close(self):
    print('Connection closed.')

#Uncomment following line if you are providing only the web socket interface and not serving static files with Tornado
application = tornado.web.Application([(r'/tsp', WSHandler),])

#Uncomment following line if you are providing both the web socket interface and static file serving with Tornado
#application = tornado.web.Application([(r'/tsp', WSHandler),(r"/content/(.*)", tornado.web.StaticFileHandler, {"path": "/var/www/visual-tsp"})])

if __name__ == "__main__":
  http_server = tornado.httpserver.HTTPServer(application)
  http_server.listen(9999)
  print('Server listening on port 9999')
  tornado.ioloop.IOLoop.instance().start()
